File: src/devices/radian.py
"""Radian Power Analyzer device driver (Python frontend).

This driver communicates with the VB.NET backend via REST API.
The backend handles actual serial port communication with Radian hardware.

Based on original VB code: original/mRadian.vb
"""

import logging
from typing import Optional
from src.devices.base import DeviceBase
from src.models.radian_metrics import RadianInstantMetrics, RadianAccumulatedMetrics, RadianPacket
from src.util.data_conversion import ti_float_to_ieee_single


class RadianDevice(DeviceBase):
    """Radian Power Analyzer device driver."""

    def __init__(self, logger: Optional[logging.Logger] = None):
        """Initialize Radian device.

        Args:
            logger: Optional logger instance
        """
        self.logger = logger or logging.getLogger(__name__)
        self._connected = False

    def connect(self, port: str, baud_rate: int = 9600) -> bool:
        """Connect to Radian power analyzer.

        Args:
            port: COM port name (e.g., "COM3")
            baud_rate: Baud rate (default 9600)

        Returns:
            True if connection successful, False otherwise
        """
        try:
            self.logger.info(f"Connecting to Radian on {port} @ {baud_rate}")
            # Connection will be handled by backend via REST API in main_window_logic
            self._connected = True
            return True
        except Exception as e:
            self.logger.error(f"Failed to connect to Radian: {e}")
            self._connected = False
            return False

    def disconnect(self) -> bool:
        """Disconnect from Radian.

        Returns:
            True if disconnection successful, False otherwise
        """
        try:
            self.logger.info("Disconnecting from Radian")
            self._connected = False
            return True
        except Exception as e:
            self.logger.error(f"Failed to disconnect from Radian: {e}")
            return False

    def is_connected(self) -> bool:
        """Check if device is connected.

        Returns:
            True if connected, False otherwise
        """
        return self._connected

    def parse_instant_metrics(self, data: bytes) -> Optional[RadianInstantMetrics]:
        """Parse instant metrics from Radian response.

        The response contains TI floating-point values for each metric.

        Args:
            data: Raw response bytes from Radian

        Returns:
            RadianInstantMetrics object or None if parsing fails
        """
        try:
            # Parse Radian packet structure
            packet = RadianPacket.from_bytes(data)
            if packet is None or len(packet.data) < 32:  # Need at least 8 floats * 4 bytes
                self.logger.error("Invalid packet or insufficient data for instant metrics")
                return None

            # Extract TI float values (4 bytes each)
            metrics = RadianInstantMetrics()
            offset = 0

            # Parse each metric (order based on original VB code)
            metrics.volt = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.amp = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.watt = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.va = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.var = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.frequency = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.phase = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.power_factor = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0

            self.logger.debug(f"Parsed instant metrics: V={metrics.volt:.2f}, "
                            f"A={metrics.amp:.2f}, W={metrics.watt:.2f}")

            return metrics

        except Exception as e:
            self.logger.error(f"Failed to parse instant metrics: {e}")
            return None

    def parse_accumulated_metrics(self, data: bytes) -> Optional[RadianAccumulatedMetrics]:
        """Parse accumulated metrics from Radian response.

        Args:
            data: Raw response bytes from Radian

        Returns:
            RadianAccumulatedMetrics object or None if parsing fails
        """
        try:
            packet = RadianPacket.from_bytes(data)
            if packet is None or len(packet.data) < 16:  # Need at least 4 floats * 4 bytes
                self.logger.error("Invalid packet or insufficient data for accumulated metrics")
                return None

            metrics = RadianAccumulatedMetrics()
            offset = 0

            # Parse accumulated energy values
            metrics.watt_hr = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.var_hr = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.q_hr = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0
            offset += 4

            metrics.va_hr = ti_float_to_ieee_single(packet.data[offset:offset+4]) or 0.0

            self.logger.debug(f"Parsed accumulated metrics: Wh={metrics.watt_hr:.2f}, "
                            f"VARh={metrics.var_hr:.2f}")

            return metrics

        except Exception as e:
            self.logger.error(f"Failed to parse accumulated metrics: {e}")
            return None

    def build_identify_command(self) -> str:
        """Build identify command.

        Returns:
            Hex string command (without checksum - backend adds it)
        """
        return "A6020000"

    def build_reset_metrics_command(self, reset_accum: bool = True,
                                   reset_instant: bool = False,
                                   reset_min: bool = False,
                                   reset_max: bool = False,
                                   reset_waveform: bool = False) -> str:
        """Build reset metrics command.

        Args:
            reset_accum: Reset accumulated metrics
            reset_instant: Reset instant metrics
            reset_min: Reset minimum values
            reset_max: Reset maximum values
            reset_waveform: Reset waveform buffers

        Returns:
            Hex string command
        """
        reset_code = 0
        if reset_waveform:
            reset_code |= 0x01
        if reset_instant:
            reset_code |= 0x02
        if reset_min:
            reset_code |= 0x04
        if reset_max:
            reset_code |= 0x08
        if reset_accum:
            reset_code |= 0x10

        return f"A6070001{reset_code:02X}"

    def build_start_accum_command(self, control_byte: int = 0x01,
                                 pulse_data: int = 0) -> str:
        """Build start accumulation command.

        Args:
            control_byte: Control mode (0x01 = normal gate start)
            pulse_data: Number of pulses (0 for query)

        Returns:
            Hex string command
        """
        pulse_msb = (pulse_data >> 8) & 0xFF
        pulse_lsb = pulse_data & 0xFF
        return f"A6080003{control_byte:02X}{pulse_msb:02X}{pulse_lsb:02X}"

    def build_start_timed_accum_command(self, time_seconds: float) -> str:
        """Build start timed accumulation test command.

        Args:
            time_seconds: Duration in seconds

        Returns:
            Hex string command
        """
        # Convert seconds to samples (20119.225 samples/sec for Rev 5+)
        samples_per_sec = 20119.225
        num_samples = int(time_seconds * samples_per_sec)

        byte0 = (num_samples >> 24) & 0xFF
        byte1 = (num_samples >> 16) & 0xFF
        byte2 = (num_samples >> 8) & 0xFF
        byte3 = num_samples & 0xFF

        return f"A60A0004{byte0:02X}{byte1:02X}{byte2:02X}{byte3:02X}"

    def build_stop_accum_command(self) -> str:
        """Build stop accumulation command.

        Returns:
            Hex string command
        """
        return "A6090000"

File: backend/SmartMeter.Backend/Devices/Radian.vb
''' <summary>
''' Radian Power Analyzer Serial Communication Driver
''' Simplified version for REST API backend
''' Based on original mRadian.vb by Frank Boudreau 2012 Landis+Gyr
''' Python port/simplification: 2026
''' </summary>

Imports System.IO.Ports
Imports System.Text
Imports System.Threading

Namespace Devices
    Public Class Radian
        Private _port As SerialPort
        Private _receiveBuffer(4095) As Byte
        Private _isConnected As Boolean = False

        Public ReadOnly Property IsConnected As Boolean
            Get
                Return _isConnected AndAlso _port IsNot Nothing AndAlso _port.IsOpen
            End Get
        End Property

        ''' <summary>
        ''' Connect to Radian on specified COM port
        ''' </summary>
        Public Sub Connect(portName As String, baudRate As Integer, Optional parity As Parity = Parity.None,
                          Optional stopBits As StopBits = StopBits.One, Optional dataBits As Integer = 8)
            Try
                If _port IsNot Nothing AndAlso _port.IsOpen Then
                    _port.Close()
                End If

                _port = New SerialPort() With {
                    .PortName = portName,
                    .BaudRate = baudRate,
                    .Parity = parity,
                    .StopBits = stopBits,
                    .DataBits = dataBits,
                    .DtrEnable = False,
                    .Handshake = Handshake.None,
                    .ReadTimeout = 2000,
                    .WriteTimeout = 2000
                }

                _port.Open()

                ' Purge any junk from buffer
                Thread.Sleep(100)
                If _port.BytesToRead > 0 Then
                    _port.DiscardInBuffer()
                End If

                _isConnected = True

            Catch ex As Exception
                _isConnected = False
                Throw New Exception($"Failed to connect to Radian on {portName}: {ex.Message}")
            End Try
        End Sub

        ''' <summary>
        ''' Disconnect from Radian
        ''' </summary>
        Public Sub Disconnect()
            Try
                If _port IsNot Nothing AndAlso _port.IsOpen Then
                    _port.Close()
                End If
                _isConnected = False
            Catch ex As Exception
                Throw New Exception($"Failed to disconnect Radian: {ex.Message}")
            End Try
        End Sub

        ''' <summary>
        ''' Send hex command string to Radian (with checksum calculation)
        ''' </summary>
        ''' <param name="hexCommand">Hex string command without checksum (e.g., "A6020000")</param>
        Public Sub SendCommand(hexCommand As String)
            Try
                If Not IsConnected Then
                    Throw New Exception("Radian not connected")
                End If

                ' Convert hex string to bytes
                Dim bytes = HexStringToBytes(hexCommand)

                ' Calculate and append 16-bit checksum
                Dim checksum As UShort = CalculateChecksum16(bytes)
                ReDim Preserve bytes(bytes.Length + 1)
                bytes(bytes.Length - 2) = CByte((checksum >> 8) And &HFF)  ' MSB
                bytes(bytes.Length - 1) = CByte(checksum And &HFF)         ' LSB

                ' Clear buffers and send
                _port.DiscardInBuffer()
                _port.DiscardOutBuffer()
                _port.Write(bytes, 0, bytes.Length)
                
                ' Debug: log sent command
                Console.WriteLine($"[Radian] Sent {bytes.Length} bytes: {BitConverter.ToString(bytes).Replace("-", "")}")

            Catch ex As Exception
                Throw New Exception($"Failed to send Radian command: {ex.Message}")
            End Try
        End Sub

        ''' <summary>
        ''' Receive response from Radian with timeout
        ''' </summary>
        ''' <param name="timeoutMs">Timeout in milliseconds</param>
        ''' <returns>Received bytes</returns>
        Public Function ReceiveResponse(Optional timeoutMs As Integer = 1000) As Byte()
            Try
                If Not IsConnected Then
                    Throw New Exception("Radian not connected")
                End If

                Dim startTime = Environment.TickCount
                Dim offset As Integer = 0
                Dim expectedLength As Integer = 0
                Dim receivedData As New List(Of Byte)

                ' Wait for data with timeout
                While (Environment.TickCount - startTime) < timeoutMs
                    If _port.BytesToRead > 0 Then
                        Dim bytesToRead = _port.BytesToRead
                        Dim buffer(bytesToRead - 1) As Byte
                        Dim bytesRead = _port.Read(buffer, 0, bytesToRead)
                        receivedData.AddRange(buffer.Take(bytesRead))
                        
                        ' Debug: log received data
                        Console.WriteLine($"[Radian] Received {bytesRead} bytes: {BitConverter.ToString(buffer, 0, bytesRead).Replace("-", "")}")

                        ' Check if we have enough data to determine packet length
                        If receivedData.Count >= 4 AndAlso expectedLength = 0 Then
                            ' Length is at bytes 2-3 (big-endian)
                            expectedLength = (CInt(receivedData(2)) << 8) Or receivedData(3)
                            expectedLength += 6  ' Add header (4 bytes) + checksum (2 bytes)
                            Console.WriteLine($"[Radian] Detected packet length: {expectedLength}, current buffer: {receivedData.Count}")
                        End If

                        ' Check if we received complete packet
                        If expectedLength > 0 AndAlso receivedData.Count >= expectedLength Then
                            Return receivedData.ToArray()
                        End If
                    Else
                        Thread.Sleep(10)
                    End If
                End While

                ' Timeout - log what we got
                Console.WriteLine($"[Radian] Timeout after {timeoutMs}ms. Received {receivedData.Count} bytes.")
                
                ' Timeout - return what we have
                If receivedData.Count > 0 Then
                    Return receivedData.ToArray()
                End If

                Throw New TimeoutException("Radian receive timeout")

            Catch ex As Exception
                Throw New Exception($"Failed to receive Radian response: {ex.Message}")
            End Try
        End Function

        ''' <summary>
        ''' Send command and wait for response
        ''' </summary>
        Public Function QueryCommand(hexCommand As String, Optional timeoutMs As Integer = 1000) As Byte()
            SendCommand(hexCommand)
            Thread.Sleep(50)  ' Small delay for device processing
            Return ReceiveResponse(timeoutMs)
        End Function

        ''' <summary>
        ''' Calculate 16-bit checksum (circular addition, modulo 65536)
        ''' </summary>
        Private Function CalculateChecksum16(data As Byte()) As UShort
            Dim checksum As Integer = 0
            For Each b As Byte In data
                checksum += b
                If checksum > 65535 Then
                    checksum -= 65536
                End If
            Next
            Return CUShort(checksum And &HFFFF)
        End Function

        ''' <summary>
        ''' Convert hex string to byte array
        ''' </summary>
        Private Function HexStringToBytes(hexString As String) As Byte()
            If hexString.Length Mod 2 <> 0 Then
                Throw New ArgumentException("Hex string length must be even")
            End If

            Dim bytes(hexString.Length \ 2 - 1) As Byte
            For i As Integer = 0 To hexString.Length - 1 Step 2
                bytes(i \ 2) = Convert.ToByte(hexString.Substring(i, 2), 16)
            Next
            Return bytes
        End Function

        ''' <summary>
        ''' Convert byte array to hex string
        ''' </summary>
        Private Function BytesToHexString(bytes As Byte()) As String
            Dim sb As New StringBuilder()
            For Each b As Byte In bytes
                sb.Append(b.ToString("X2"))
            Next
            Return sb.ToString()
        End Function

        ' ===== RADIAN COMMAND CONSTANTS =====

        Public Const CMD_IDENTIFY As String = "A6020000"
        Public Const CMD_RESET_RD As String = "A6030000"
        Public Const CMD_STOP_ACCUM As String = "A6090000"
        Public Const CMD_RESET_METRICS_BASE As String = "A6070001"
        Public Const CMD_START_ACCUM_BASE As String = "A6080003"
        Public Const CMD_START_TIMED_ACCUM_BASE As String = "A60A0004"

        ''' <summary>
        ''' Build Reset Metrics command
        ''' </summary>
        Public Function BuildResetMetricsCommand(resetWaveform As Boolean, resetInstant As Boolean,
                                                 resetInstantMin As Boolean, resetInstantMax As Boolean,
                                                 resetAccum As Boolean) As String
            Dim resetCode As Byte = 0

            If resetWaveform Then resetCode = resetCode Or &H1
            If resetInstant Then resetCode = resetCode Or &H2
            If resetInstantMin Then resetCode = resetCode Or &H4
            If resetInstantMax Then resetCode = resetCode Or &H8
            If resetAccum Then resetCode = resetCode Or &H10

            Return CMD_RESET_METRICS_BASE & resetCode.ToString("X2")
        End Function

        ''' <summary>
        ''' Build Start Accumulation command
        ''' </summary>
        Public Function BuildStartAccumCommand(controlByte As Byte, pulseData As UShort) As String
            Dim pulseMSB As Byte = CByte((pulseData >> 8) And &HFF)
            Dim pulseLSB As Byte = CByte(pulseData And &HFF)

            Return CMD_START_ACCUM_BASE & controlByte.ToString("X2") &
                   pulseMSB.ToString("X2") & pulseLSB.ToString("X2")
        End Function

        ''' <summary>
        ''' Build Start Timed Accumulation Test command
        ''' </summary>
        Public Function BuildStartTimedAccumCommand(timeInSeconds As Single) As String
            ' Convert seconds to samples (20119.225 samples/sec for Rev 5+)
            Dim samplesPerSecond As Single = 20119.225
            Dim numberOfSamples As UInteger = CUInt(timeInSeconds * samplesPerSecond)

            Dim byte0 As Byte = CByte((numberOfSamples >> 24) And &HFF)
            Dim byte1 As Byte = CByte((numberOfSamples >> 16) And &HFF)
            Dim byte2 As Byte = CByte((numberOfSamples >> 8) And &HFF)
            Dim byte3 As Byte = CByte(numberOfSamples And &HFF)

            Return CMD_START_TIMED_ACCUM_BASE & byte0.ToString("X2") &
                   byte1.ToString("X2") & byte2.ToString("X2") & byte3.ToString("X2")
        End Function
    End Class
End Namespace
